dask/dataframe/core.py

```diff
@@ -3519,7 +3519,16 @@ def apply_and_enforce(func, args, kwargs, meta):
     if isinstance(df, (pd.DataFrame, pd.Series, pd.Index)):
         if len(df) == 0:
             return meta
-        c = meta.columns if isinstance(df, pd.DataFrame) else meta.name
+
+        if isinstance(df, pd.DataFrame):
+            # Need nan_to_num otherwise nan comparison gives False
+            if not np.array_equal(np.nan_to_num(meta.columns),
```

Comment:
Does `if not meta.columns.equals(df.columns)` work? That seems to handle `NaN` well.

Would it be possible to include the mismatched columns, perhaps if they aren't too large?

Reply:
Thanks for your comment, I was not aware of `.equals` but it seems to do the trick!

Re: including the mismatched columns, do you mean including what columns are expected based on the metadata, and what columns are actually present? So something like `expected 'date' from metadata, found 'ent'` for each mismatch?

Reply:
Yeah I was thinking something like that, but it can be tricky if the output is large. I think just improving the error message based on @mrocklin's [suggestion](https://github.com/dask/dask/pull/3485#discussion_r187301386) will be sufficient.